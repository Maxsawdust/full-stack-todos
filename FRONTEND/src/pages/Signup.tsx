import { useFormik } from "formik";
import {
  FormInput,
  FormSubmitButton,
  Heading,
  SubHeading,
} from "../components";
import { signupSchema } from "../utils/validationSchema";
import { useNavigate } from "react-router";

export default function Signup() {
  const navigate = useNavigate();

  const formik = useFormik({
    initialValues: {
      username: "",
      email: "",
      password: "",
      confirmPassword: "",
    },

    onSubmit: () => {
      // register new user
      // navigate to login
      navigate("/login");
    },

    validateOnBlur: false,
    validateOnChange: false,

    validationSchema: signupSchema,
  });

  return (
    <div className="page">
      <div className="w-80 h-3/5 flex flex-col gap-6">
        <header className="">
          <Heading>Sign Up</Heading>
          <SubHeading>Productivity starts here</SubHeading>
        </header>

        <div className="h-[2px] bg-gray-400" />

        <form className="flex flex-col gap-4" onSubmit={formik.handleSubmit}>
          <FormInput
            errorMessage={formik.errors.username}
            onChange={formik.handleChange}
            name="username">
            Username
          </FormInput>
          <FormInput
            errorMessage={formik.errors.email}
            onChange={formik.handleChange}
            name="email">
            Email
          </FormInput>
          <FormInput
            errorMessage={formik.errors.password}
            onChange={formik.handleChange}
            name="password">
            Password
          </FormInput>
          <FormInput
            errorMessage={formik.errors.confirmPassword}
            onChange={formik.handleChange}
            name="confirmPassword">
            Confirm Password
          </FormInput>
          <FormSubmitButton>Sign Up</FormSubmitButton>
        </form>
      </div>
    </div>
  );
}
