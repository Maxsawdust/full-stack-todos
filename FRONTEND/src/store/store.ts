import { configureStore } from "@reduxjs/toolkit";
import loadingReducer from "./reducers/loadingReducer";

const store = configureStore({
  reducer: {
    loadingReducer: loadingReducer,
  },
});

export default store;

/* as referenced here: https://react-redux.js.org/using-react-redux/usage-with-typescript */
// type for the state of the store
export type RootState = ReturnType<typeof store.getState>;
// type for the dispatch of the store
export type AppDispatch = typeof store.dispatch;
