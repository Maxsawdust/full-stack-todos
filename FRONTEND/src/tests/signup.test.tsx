import { render, screen, waitFor } from "@testing-library/react";
import Signup from "../pages/Signup";
import { Provider } from "react-redux";
import store from "../store/store";
import userEvent from "@testing-library/user-event";

// Mock the useNavigate hook from react-router
// This tracks when navigation occurs without actually navigating
const mockNavigate = vi.fn();
vi.mock("react-router", () => ({
  ...vi.importActual("react-router"),
  useNavigate: () => mockNavigate,
}));

// Mock fetch
const mockFetch = vi.fn();
window.fetch = mockFetch;

// Mock Redux store
const mockDispatch = vi.fn();
vi.mock("../store/hooks/hooks", () => ({
  useAppDispatch: () => mockDispatch,
  useAppSelector: () => ({ isLoading: false }),
}));

describe("Signup Tests", () => {
  // Helper function to set up the test environment
  const setup = () => {
    // Configure fetch mock to return a successful response
    mockFetch.mockResolvedValue({
      ok: true,
      json: () => Promise.resolve({ message: "Success" }),
    });

    // Render the Signup component with necessary providers
    render(
      <Provider store={store}>
        <Signup />
      </Provider>
    );
  };

  // Test that the signup page renders correctly
  it("should render the signup page", () => {
    setup();
    // Compare the rendered component to a snapshot
    expect(<Signup />).toMatchSnapshot();
  });

  it("should navigate to login on singup", async () => {
    setup();

    // Get form elements by their placeholder text
    const usernameField = screen.getByPlaceholderText("Username");
    const emailField = screen.getByPlaceholderText("Email");
    const passwordField = screen.getByPlaceholderText("Password");
    const confirmPasswordField =
      screen.getByPlaceholderText("Confirm Password");
    const signupButton = screen.getByTestId("Sign Up");

    // Fill in the form with valid data
    await userEvent.type(usernameField, "testuser");
    await userEvent.type(emailField, "testuser@example.com");
    await userEvent.type(passwordField, "Password1!");
    await userEvent.type(confirmPasswordField, "Password1!");

    // Submit the form
    await userEvent.click(signupButton);

    // Verify that the fetch API was called with the correct parameters
    await waitFor(() => {
      expect(mockFetch).toHaveBeenCalledWith(
        "http://localhost:5000/users/signup",
        expect.objectContaining({
          method: "POST",
          headers: { "Content-Type": "application/json" },
          body: expect.any(String),
        })
      );
    });

    // Verify that navigation to the login page occurred
    await waitFor(() => {
      expect(mockNavigate).toHaveBeenCalledWith("/login");
    });
  });
});
