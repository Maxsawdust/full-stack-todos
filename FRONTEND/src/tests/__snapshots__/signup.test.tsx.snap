// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Signup Tests > should render the signup page 1`] = `<Signup />`;
